### PR TITLE
Add mypy linting

### PR DESCRIPTION
**Pull request by MartinBasti**
* add mypy to tox
* add mypy to GH actions
* update missing definitions
* make gating.yaml more readable by using matrix for linters

Signed-off-by: Martin Basti <mbasti@redhat.com>

### CODE REVIEW

**Comment by MartinBasti** [comment]
Pro tips how to address these, welcome
```
cachito/common/paths.py:14: error: Unsupported dynamic base class "type"
cachito/workers/paths.py:38: error: Unsupported dynamic base class "type"
cachito/workers/pkg_managers/gomod.py:100: error: Argument "key" to "max" has incompatible type "Callable[[Sized], int]"; expected "Callable[[Optional[str]], SupportsLessThan]"
cachito/workers/pkg_managers/general_js.py:352: error: Incompatible types in assignment (expression has type "object", variable has type "str")

```

**Review by tkdchen — APPROVED** [review]
Really NICE!

File: cachito/workers/pkg_managers/pip.py [hunk 1]
```diff
@@ -8,6 +8,7 @@
 import secrets
 import shutil
 import tarfile
+from typing import Optional
```
Comment (serg-cymbaluk) [inline cachito/workers/pkg_managers/pip.py:11]: BTW are we going to add isort check?

Comment (akhmanova) [reply cachito/workers/pkg_managers/pip.py:11]: sounds good, but how do you want to organize both `black` and `isort`?

File: .github/workflows/gating.yaml [hunk 2]
```diff
@@ -70,40 +78,11 @@ jobs:
...
-      with:
-        python-version: 3.8
-    - name: Install dependencies
-      run: |
-        python -m pip install --upgrade pip
-        pip install tox
-    - name: Test with tox
-      run: tox -e black
+    - name: Test '${{ matrix.tox_env }}' with tox
+      run: tox -e ${{ matrix.tox_env }}
```
Comment (akhmanova) [inline .github/workflows/gating.yaml:83]: Really good, thank you!

**Review by akhmanova — APPROVED** [review]

File: cachito/workers/pkg_managers/pip.py (hunk 1 above)
Comment (MartinBasti) [reply cachito/workers/pkg_managers/pip.py:11]: I don't have such plan, I never used isort. I just remember that fiting pylint import order requirements was a pain, sometime impossible :D

Comment (MartinBasti) [reply cachito/workers/pkg_managers/pip.py:11]: But feel free to add it

**Pull request merged**